cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+ * @tparam PREFER_POW2_BITS Whether the number of bits to dedicate to each counter should be a power-of-two. If enabled,
+ * this allows replacing integer multiplication with a bit-shift in exchange for higher register pressure.
+ * @tparam BackingUnitT The data type that is used to provide the bits of all the counters that shall be allocated.
+ */
+template <uint32_t NUM_ITEMS, uint32_t MAX_ITEM_VALUE, bool PREFER_POW2_BITS, typename BackingUnitT = uint32_t>
+class BitPackedCounter
+{
+private:
+  /**
+   * @brief Returns a bit-mask that has the \p num_bits least-significant bits being set.
+   *
+   * @param num_bits The number of least-significant bits to set to '1'
+   */
+  static __forceinline__ __host__ __device__ uint32_t GET_BIT_MASK(uint32_t num_bits)
+  {
+    return (num_bits >= (8U * sizeof(uint32_t))) ? Traits<uint32_t>::MAX_KEY : (0x01U << num_bits) - 1;
+  }
+
+  /**
+   * @brief Returns a 64-bit unsigned integer that has the \p num_bits least-significant bits being set.
+   *
+   * @param num_bits The number of least-significant bits to set to '1'
+   */
+  static __forceinline__ __host__ __device__ uint64_t GET_BIT_MASKLL(uint32_t num_bits)
+  {
+    return (num_bits >= (8U * sizeof(uint64_t))) ? Traits<uint64_t>::MAX_KEY : (0x01ULL << num_bits) - 1ULL;
+  }
+
+  enum : uint32_t
+  {
+    /// The minimum number of bits required to represent all values from [0, MAX_ITEM_VALUE]
+    MIN_BITS_PER_ITEM = MAX_ITEM_VALUE == 0U ? 1U : cub::Log2<(MAX_ITEM_VALUE + 1)>::VALUE,
+
+    /// The number of btis allocated for each item. For pre-Volta, we prefer a power-of-2 here to have the compiler
+    /// replace costly integer multiplication with bit-shifting.
+    BITS_PER_ITEM = PREFER_POW2_BITS ? (0x01ULL << (cub::Log2<MIN_BITS_PER_ITEM>::VALUE)) : MIN_BITS_PER_ITEM,
+
+    /// The number of bits that each backing data type can store
+    NUM_BITS_PER_UNIT = sizeof(BackingUnitT) * 8,
+
+    /// The number of items that each backing data type can store
+    ITEMS_PER_UNIT = NUM_BITS_PER_UNIT / BITS_PER_ITEM,
+
+    /// The number of bits the backing data type is actually making use of
+    USED_BITS_PER_UNIT = ITEMS_PER_UNIT * BITS_PER_ITEM,
+
+    /// The number of backing data types required to store the given number of items
+    NUM_TOTAL_UNITS = CUB_QUOTIENT_CEILING(NUM_ITEMS, ITEMS_PER_UNIT),
+
+    /// This is the net number of bit-storage provided by each unit (remainder bits are unused)
+    UNIT_MASK = (USED_BITS_PER_UNIT >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << USED_BITS_PER_UNIT) - 1,
+    /// This is the bit-mask for each item
+    ITEM_MASK = (BITS_PER_ITEM >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << BITS_PER_ITEM) - 1
+  };
+
+  //------------------------------------------------------------------------------
+  // ACCESSORS
+  //------------------------------------------------------------------------------
+public:
+  __device__ __forceinline__ uint32_t Get(uint32_t index) const
```

Comment:
Could all those methods be `noexcept`?

Reply:
As I understand, there are no exceptions in CUDA, so I'm not sure what this would change.